src/wayland/shell/legacy/decoration.rs

```diff
@@ -0,0 +1,128 @@
+//! use wayland_protocols::misc::server_decoration::server::org_kde_kwin_server_decoration_manager::Mode;
+//!
+//! # let mut display = wayland_server::Display::new();
+//!
+//! let decoration_manager = init_kde_decoration_manager(
+//!     &mut display,
+//!     |decoration, request| match request {
+//!         KdeDecorationRequest::Setup { manager } => manager.default_mode(Mode::Server),
+//!         KdeDecorationRequest::NewToplevelDecoration { surface } => (),
+//!         KdeDecorationRequest::RequestMode { mode, .. } => decoration.mode(mode),
+//!         KdeDecorationRequest::Release => (),
+//!     },
+//!     None,
+//! );
+
+use std::{cell::RefCell, ops::Deref, rc::Rc};
+
+use wayland_protocols::misc::server_decoration::server::{
+    org_kde_kwin_server_decoration::{Mode, OrgKdeKwinServerDecoration, Request as ServerDecorationRequest},
+    org_kde_kwin_server_decoration_manager::{
+        OrgKdeKwinServerDecorationManager, Request as ServerDecorationManagerRequest,
+    },
+};
+use wayland_server::protocol::wl_surface::WlSurface;
+use wayland_server::{Display, Filter, Global, Main};
+
+/// Events generated by the kde decoration manager.
+#[derive(Debug)]
+pub enum KdeDecorationRequest {
+    /// Decoration manager setup completion.
+    Setup {
+        /// KDE decoration manager.
+        manager: OrgKdeKwinServerDecorationManager,
+    },
+    /// A new toplevel decoration was instantiated.
+    NewToplevelDecoration {
+        /// KDE decoration object.
+        decoration: OrgKdeKwinServerDecoration,
+        /// The surface asosiated with the decoration.
+        surface: WlSurface,
+    },
+    /// Informs the compositor that the client prefers the provided decoration mode.
+    RequestMode {
+        /// KDE decoration object.
+        decoration: OrgKdeKwinServerDecoration,
+        /// The surface asosiated with the decoration.
+        surface: WlSurface,
+        /// Desired mode.
+        mode: Mode,
+    },
+    /// Decoration object was released.
+    Release {
+        /// The surface asosiated with the decoration.
+        surface: WlSurface,
+    },
+}
+
+/// Create a new KDE Decoration Manager global
+///
+/// A compositor can set the default mode for new clients at any time using
+/// [`OrgKdeKwinServerDecorationManager::default_mode`]. For individual windows the decorations can
+/// be set using [`OrgKdeKwinServerDecoration::mode`]. Both of these requests will cause the client
+/// to respond with its desired mode using [`KdeDecorationRequest::RequestMode`], which then should
+/// be acknowledged by the compositor using another [`OrgKdeKwinServerDecoration::mode`] event.
+/// **It is up to the compositor to prevent feedback loops**, a client is free to ignore the
+/// suggested decoration mode.
+pub fn init_kde_decoration_manager<L, Impl>(
+    display: &mut Display,
+    implementation: Impl,
+    _logger: L,
```

Comment:
I think this logger deserves a little use, it's trying so hard to be there!

I guess some trace or debug level messages would be fine for some debugging if needed.

Reply:
Just copied this from the XDG implementation really and thought the same. I'd assume there it at least was there to future-proof the protocol but assuming this gets merged without bugs there likely won't be any future changes that would require a logger.

Just adding some debug output might make sense, otherwise there's not much reason to have it.

Reply:
Maybe log default_mode changes (if implemented)?

Reply:
Would depend on *how* it is implemented. If downstream has direct access to the `OrgKdeKwinServerDecorationManager` then default mode can be changed without smithay ever getting any chance to log that. Though the default mode set at startup could still be logged.